src/components/Barometer.tsx

```diff
@@ -0,0 +1,181 @@
+ * the safe area insets. We add these to the padding.
+ *
+ * Finally, we keep track of the keyboard, adding extra padding &
+ * scheduling animations as needed.
+ */
+export function Barometer(props: Props): JSX.Element {
+  const { children, onLayout = () => {} } = props
+
+  // Mutable state:
+  const keyboardHeight: React.MutableRefObject<number> = React.useRef(0)
+  const lastLayoutJson: React.MutableRefObject<string> = React.useRef('')
+  const view: React.RefObject<SafeAreaView | View> = React.useRef(null)
+  const childView: React.RefObject<View> = React.useRef(null)
+
+  // Handle layout changes:
+  const handleLayout = React.useCallback((): void => {
+    // Measure the view in the window:
+    const viewPromise: Promise<SideList> = new Promise(resolve => {
+      if (view.current == null) return resolve(emptySides)
+      view.current.measureInWindow((x, y, width, height) => {
+        const window = Dimensions.get('window')
+        resolve([y, window.width - width - x, window.height - height - y, x])
+      })
+    })
+
+    // Measure the child view in the window:
+    const childPromise: Promise<SideList> = new Promise(resolve => {
+      if (childView.current == null) return resolve(viewPromise)
+      childView.current.measureInWindow((x, y, width, height) => {
+        const window = Dimensions.get('window')
+        resolve([y, window.width - width - x, window.height - height - y, x])
+      })
+    })
+
+    // Measure the gap between the bottom of the screen and the view:
+    const bottomPromise: Promise<number> = new Promise(resolve => {
+      if (view.current == null) return 0
+      view.current.measure((x, y, width, height, screenX, screenY) => {
+        const screen = Dimensions.get('screen')
+        resolve(screen.height - height - screenY)
+      })
+    })
+
+    // Combine the results, then call the callback:
+    Promise.all([viewPromise, childPromise, bottomPromise])
+      .then(([viewOffset, childOffset, bottomGap]) => {
+        // Cancel out any offset, so we cover the full window:
+        const offset = mapSides(viewOffset, side => -Math.max(side, 0))
+
+        // If the offset is negative, issue positive padding,
+        // plus any safe area:
+        const safePadding = subtractSides(childOffset, viewOffset)
+        const padding = addSides(
+          safePadding,
+          mapSides(viewOffset, side => Math.abs(side))
+        )
+
+        // Use the keyboard padding, if needed:
+        const keyboardPadding = Math.max(
+          keyboardHeight.current - bottomGap - offset[2],
+          0
+        )
+        padding[2] = Math.max(padding[2], keyboardPadding)
+
+        // Send an update if we have changes:
+        const string = JSON.stringify([offset, padding])
+        if (string !== lastLayoutJson.current) {
+          lastLayoutJson.current = string
+          onLayout({
+            offset: sidesToOffset(offset),
+            padding: sidesToPadding(padding)
+          })
+        }
```

Comment:
Approved with optional changes here to include `offsetGaps` (or similarly named) measurements in place of `bottomGap` to account for known top-gap on Android devices. The idea is to include these offset gaps when calculating the offset and padding outputs.